diff --git a/voxen.c b/voxen.c
index 4d37e69..3aaeaed 100644
--- a/voxen.c
+++ b/voxen.c
@@ -222,7 +222,7 @@ float statusTextDecayFinished = 0.0f;
 // ----------------------------------------------------------------------------
 // Lights
 // Could reduce spotAng to minimal bits.  I only have 6 spot lights and half are 151.7 and other half are 135.
-GLuint lightsID, voxelLightListIndicesID, voxelLightListsRawID, lightShadowsEnabledID, shadowMapsIndirectionID;
+GLuint lightsID, voxelLightListIndicesID, voxelLightListsRawID, shadowMapsIndirectionID;
 uint32_t* voxelLightListsRaw = NULL;
 uint32_t* voxelLightListIndices = NULL;
 uint32_t* shadowmapIndirectionList = NULL;
@@ -230,9 +230,9 @@ uint16_t numDynamicLights;
 float lights[LIGHT_COUNT * LIGHT_DATA_SIZE] = {0};
 float lightsRangeSquared[LIGHT_COUNT] = {0.0f};
 bool lightDirty[LIGHT_COUNT] = { [0 ... LIGHT_COUNT-1] = true };
-float*** lightViewProj = NULL; // Array of Array of 6 Arrays of 16 floats (matrix 4x4).  lightViewProj[i][face][0 ... 15]
-float*** lightView = NULL; // Array of Array of 6 Arrays of 16 floats (matrix 4x4).  lightView[i][face][0 ... 15]
-FrustumPlane*** lightFrustumPlanes = NULL; // Array of Array of 6 Arrays of FrustumPlane structs (four floats).  lightFrustumPlanes[i][face][.nx,.ny,, .nz, .d]
+float (*lightView)[6][4][4] = NULL; // Array of Array of 6 Arrays of 16 floats (matrix 4x4).  lightView[i][face][0 ... 15]
+float (*lightViewProj)[6][4][4] = NULL; // Array of Array of 6 Arrays of 16 floats (matrix 4x4).  lightViewProj[i][face][0 ... 15]
+FrustumPlane (*lightFrustumPlanes)[6][6] = NULL; // Array of Array of 6 Arrays of FrustumPlane structs (four floats).  lightFrustumPlanes[i][face][.nx,.ny,, .nz, .d]
 // ----------------------------------------------------------------------------
 // ============================================================================
 // OpenGL / Rendering Helper Functions
@@ -653,8 +653,6 @@ void UpdateVoxelLightLists() {
     glNamedBufferData(lightsID,loadedLights * LIGHT_DATA_SIZE * sizeof(float), lights, GL_DYNAMIC_DRAW);
 }
 
-uint32_t* lightShadowsEnabled = NULL;
-
 void VoxelLists() {
     DebugRAM("start of VoxelLists");
     voxelLightListsRaw = malloc(VOXEL_COUNT * 4 * sizeof(uint32_t));
@@ -663,34 +661,18 @@ void VoxelLists() {
     voxelLightListsRawID = SetupSSBO(voxelLightListsRawID, 27,  1008105 * sizeof(uint32_t), NULL, GL_DYNAMIC_DRAW);
     shadowmapIndirectionList = malloc(loadedLights * sizeof(uint32_t));
     lightsID = SetupSSBO(lightsID, 19, loadedLights * LIGHT_DATA_SIZE * sizeof(float), NULL, GL_DYNAMIC_DRAW);
-    lightView = malloc(loadedLights * sizeof(float**));
-    lightViewProj = malloc(loadedLights * sizeof(float**));
-    lightFrustumPlanes = malloc(loadedLights * sizeof(FrustumPlane**));
-    for (int i=0;i<loadedLights;++i) {
-        lightView[i] = malloc(6 * sizeof(float*));
-        lightViewProj[i] = malloc(6 * sizeof(float*));
-        lightFrustumPlanes[i] = malloc(6 * sizeof(FrustumPlane*));
-        for (int j=0;j<6;++j) {
-            lightView[i][j] = malloc(4 * 4 * sizeof(float)); // Matrix 4x4 for this cubemap face
-            lightViewProj[i][j] = malloc(4 * 4 * sizeof(float)); // Matrix 4x4 for this cubemap face
-            lightFrustumPlanes[i][j] = malloc(6 * sizeof(FrustumPlane)); // Frustum Planes for this cubemap face
-        }
-    }
-    
+    lightView      = calloc(loadedLights * 6 * 4 * 4, sizeof(float));
+    lightViewProj  = calloc(loadedLights * 6 * 4 * 4, sizeof(float));
+    lightFrustumPlanes   = calloc(loadedLights * 6 * 6, sizeof(FrustumPlane));
     DebugRAM("prior to UpdateVoxelLightLists");
     UpdateVoxelLightLists();
     for (uint16_t i = 3; i < loadedInstances; i++) UpdateInstanceMatrix(i); // Skip player indices and start at 3
     matricesBuffer = SetupSSBO(matricesBuffer, 11, loadedInstances * 16 * sizeof(float), modelMatrices, GL_DYNAMIC_DRAW);
-    lightShadowsEnabled = malloc(loadedLights * sizeof(uint32_t));
-    memset(lightShadowsEnabled,1u,loadedLights * sizeof(uint32_t));
-    lightShadowsEnabledID = SetupSSBO(lightShadowsEnabledID, 6, loadedLights * sizeof(uint32_t), lightShadowsEnabled, GL_STATIC_DRAW);
     shadowMapsIndirectionID = SetupSSBO(shadowMapsIndirectionID, 8, loadedLights * sizeof(uint32_t), NULL, GL_STATIC_DRAW);
     DebugRAM("end of VoxelLists");
 }
 
 void RenderShadowmap(uint16_t lightIdx) {
-    if (lightShadowsEnabled[lightIdx] == 0u) return;
-    
     uint32_t litIdx = lightIdx * LIGHT_DATA_SIZE;
     float lightPosX = lights[litIdx + LIGHT_DATA_OFFSET_POSX];
     float lightPosY = lights[litIdx + LIGHT_DATA_OFFSET_POSY];
@@ -770,6 +752,7 @@ void RenderShadowmap(uint16_t lightIdx) {
 }
 
 void RenderShadowmaps(void) {
+    DebugRAM("start of RenderShadowmaps");
     if (settings_Shadows < 1u) return;
 
     glUseProgram(shadowmapsClearShaderProgram);
@@ -788,14 +771,12 @@ void RenderShadowmaps(void) {
     glDepthMask(GL_TRUE);
     glBindVertexArray(vao_chunk);
     
-    // This sort method is 0.027ms so not too worried about it at the moment.
+    // This sort method is 0.027ms... now that we do this whenever changing world cells, so like 0 to 3 times per second, it's rather worrisome.
 //     double sortStart = get_time();
     // Collect candidates: only lights that are enabled, within FAR_PLANE, and in PVS
     LightCandidate candidates[loadedLights];
     uint32_t candidateCount = 0;
     for (uint16_t i = 0; i < loadedLights; ++i) {
-        if (lightShadowsEnabled[i] == 0u) continue;
-
         uint32_t litIdx = i * LIGHT_DATA_SIZE;
         float lightPosX = lights[litIdx + LIGHT_DATA_OFFSET_POSX];
         float lightPosY = lights[litIdx + LIGHT_DATA_OFFSET_POSY];
