diff --git a/entity.c b/entity.c
index 4086dcd..af79ad3 100644
--- a/entity.c
+++ b/entity.c
@@ -3,6 +3,7 @@
 #include <stdio.h>
 #include <errno.h>
 #include "entity.h"
+#include "event.h"
 #include "voxen.h"
 #include "vmath.h"
 
@@ -41,7 +42,6 @@ void GetLevel_Transform_Offsets(int32_t curlevel, float* ofsx, float* ofsy, floa
     switch(curlevel) { // Match the parent transforms #.NAMELevel, e.g. 1.MedicalLevel
         case 0:  *ofsx = 3.6f; *ofsy = -4.10195f; *ofsz = 1.0f; break;
         case 1:  *ofsx = 25.56f; *ofsy = -48.64f; *ofsz = -5.2f; break;
-//         case 1:  *ofsx = 0.0f; *ofsy = 0.0f; *ofsz = 0.0f; break;
         case 2:  *ofsx = -2.6f; *ofsy = 0.0f; *ofsz = -7.7f; break;
         case 3:  *ofsx = -45.12f; *ofsy = -0.700374f; *ofsz = -16.32f; break;
         case 4:  *ofsx = -20.4f; *ofsy = 0.0f; *ofsz = 11.48f; break;
@@ -387,6 +387,34 @@ void CopyInstanceRegion(uint16_t head, uint16_t* instanceTypeArray, Entity* temp
     }
 }
 
+// Using "relative time" = pauseRelativeTime and "finished" = some
+// script's timer float value, e.g. attackFinished, in the notes below...
+//
+// If the relative time is 123 when we save and finished is 156, then when
+// we load and relative time is 160 and we set finished to 156, it will
+// immediately finish.
+//
+// Need to take finished - relative time = 33 and save that.  Then on load
+// take this differential value and do relative time + differential = 160 +
+// 33 = 193, then the same condition is restored such that the finished 
+// timer still has 33 before it is up.
+//
+// In the scenario where finished is less than relative time, if finished =
+// 103 and relative time is still 123, then when we load and relative time
+// is 160, all is still fine, timer is already up.
+//
+// Still can't hurt to do finished - relative time = -20. Then when we load
+// the value do relative time 160 + -20 = 140.  This is arguably best just
+// in case there is a whackado one-off instance of comparing (time - 
+// finished) somewhere instead of (finished < time) which is my usual Quake
+// derived timer pattern.
+float LoadRelativeTimeDifferential(char* trimmed_value, char* initialLine, uint32_t lineNum) {
+    return parse_float(trimmed_value, initialLine, lineNum) + pauseRelativeTime; // Add current instance's relative time to get same timer in context of current time.  See above notes.
+}
+
+uint16_t lightsIndirection[INSTANCE_COUNT];
+uint16_t firstRelevantLightsIndirectionIndex = 0;
+
 void SortInstances(void) { // Reorder instances such that each type is grouped opaque->doublesided->transparent in that order in instances[].
     double start_time = get_time();
     DualLog("Sorting entity instances... ");
@@ -468,45 +496,68 @@ void SortInstances(void) { // Reorder instances such that each type is grouped o
             instances[i].volume = entities[entIdx].volume * 0.5f;
         }
     }
+    
+    memset(lightsIndirection,UINT16_MAX,INSTANCE_COUNT * sizeof(uint16_t));
+    uint16_t lightsHead = 0;
+    bool foundALightYet = false;
+    for (int i = 0; i < INSTANCE_COUNT; ++i) { // Start at 0 since ANYTHING could be a light, e.g. projectiles
+        if (!(instances[i].entflags & ENTFLAG_IS_LIGHT)) continue;
+        
+        if (!foundALightYet){
+            firstRelevantLightsIndirectionIndex = i;
+            foundALightYet = true;
+        }
+        
+        lightsIndirection[i] = lightsHead;
+        lightsHead++;
+    }
 }
 
 void AddInstance(uint16_t entIdx, uint16_t instanceIdx, uint32_t lineNum) {
-    if (entIdx >= entityCount) { DualLogError("\nEntity index when loading level geometry object %d was %d, exceeds max defined entity count of %d\n",lineNum,entIdx,entityCount); exit(1); }
-            
     instances[instanceIdx].index = entIdx;
-    instances[instanceIdx].modelIndex = entities[entIdx].modelIndex;
-    if (instances[instanceIdx].modelIndex < loadedModels) renderableCount++;
-    instances[instanceIdx].texIndex = entities[entIdx].texIndex;
-    instances[instanceIdx].glowIndex = entities[entIdx].glowIndex;
-    if (instances[instanceIdx].glowIndex >= MATERIAL_IDX_MAX) instances[instanceIdx].glowIndex = 0;
-    instances[instanceIdx].specIndex = entities[entIdx].specIndex;
-    if (instances[instanceIdx].specIndex >= MATERIAL_IDX_MAX) instances[instanceIdx].specIndex = 0;
-    instances[instanceIdx].normIndex = entities[entIdx].normIndex;
-    if (instances[instanceIdx].normIndex >= MATERIAL_IDX_MAX) instances[instanceIdx].normIndex = 0;
-    instances[instanceIdx].lodIndex = entities[entIdx].lodIndex;
-//     instances[instanceIdx].entflags = entities[entIdx].entflags; // Decided this was dangerous/error-prone, commented out in lieu of these explicit sets to better preserve the loaded data:
-    flag_set(&instances[instanceIdx].entflags, ENTFLAG_CARDCHUNK,  entities[entIdx].entflags & ENTFLAG_CARDCHUNK);
-    flag_set(&instances[instanceIdx].entflags, ENTFLAG_USEGRAVITY,  entities[entIdx].entflags & ENTFLAG_USEGRAVITY);
-    flag_set(&instances[instanceIdx].entflags, ENTFLAG_KINEMATIC,  entities[entIdx].entflags & ENTFLAG_KINEMATIC);
-    flag_set(&instances[instanceIdx].entflags, ENTFLAG_RIGIDBODY,  entities[entIdx].entflags & ENTFLAG_RIGIDBODY);
-    instances[instanceIdx].collider = entities[entIdx].collider;
-    instances[instanceIdx].colliderCenter.x = entities[entIdx].colliderCenter.x;
-    instances[instanceIdx].colliderCenter.y = entities[entIdx].colliderCenter.y;
-    instances[instanceIdx].colliderCenter.z = entities[entIdx].colliderCenter.z;
-    instances[instanceIdx].colliderSize.x = entities[entIdx].colliderSize.x;
-    instances[instanceIdx].colliderSize.y = entities[entIdx].colliderSize.y;
-    instances[instanceIdx].colliderSize.z = entities[entIdx].colliderSize.z;
-    instances[instanceIdx].colliderMeshIndex = entities[entIdx].colliderMeshIndex;
-    instances[instanceIdx].mass = entities[entIdx].mass > 0.0f ? entities[entIdx].mass : 1.0f; // Nonzero fallback.
-    instances[instanceIdx].linearDrag = entities[entIdx].linearDrag > 0.0f ? entities[entIdx].linearDrag : 0.0f;
-    instances[instanceIdx].angularDrag = entities[entIdx].angularDrag > 0.0f ? entities[entIdx].angularDrag : 0.05f;
+    bool isLight = (instances[instanceIdx].entflags & ENTFLAG_IS_LIGHT);
+    if (entIdx >= entityCount) {
+        if (!isLight) { DualLogError("\nEntity index when loading non-light entity %d was %d, exceeds max defined entity count of %d\n",lineNum,entIdx,entityCount); exit(1); }
+    } else {
+        instances[instanceIdx].modelIndex = entities[entIdx].modelIndex;
+        if (instances[instanceIdx].modelIndex < loadedModels) renderableCount++;
+        instances[instanceIdx].texIndex = entities[entIdx].texIndex;
+        instances[instanceIdx].glowIndex = entities[entIdx].glowIndex;
+        if (instances[instanceIdx].glowIndex >= MATERIAL_IDX_MAX) instances[instanceIdx].glowIndex = 0;
+        instances[instanceIdx].specIndex = entities[entIdx].specIndex;
+        if (instances[instanceIdx].specIndex >= MATERIAL_IDX_MAX) instances[instanceIdx].specIndex = 0;
+        instances[instanceIdx].normIndex = entities[entIdx].normIndex;
+        if (instances[instanceIdx].normIndex >= MATERIAL_IDX_MAX) instances[instanceIdx].normIndex = 0;
+        instances[instanceIdx].lodIndex = entities[entIdx].lodIndex;
+    //     instances[instanceIdx].entflags = entities[entIdx].entflags; // Decided this was dangerous/error-prone, commented out in lieu of these explicit sets to better preserve the loaded data:
+        flag_set(&instances[instanceIdx].entflags, ENTFLAG_CARDCHUNK,  entities[entIdx].entflags & ENTFLAG_CARDCHUNK);
+        flag_set(&instances[instanceIdx].entflags, ENTFLAG_USEGRAVITY,  entities[entIdx].entflags & ENTFLAG_USEGRAVITY);
+        flag_set(&instances[instanceIdx].entflags, ENTFLAG_KINEMATIC,  entities[entIdx].entflags & ENTFLAG_KINEMATIC);
+        flag_set(&instances[instanceIdx].entflags, ENTFLAG_RIGIDBODY,  entities[entIdx].entflags & ENTFLAG_RIGIDBODY);
+        instances[instanceIdx].collider = entities[entIdx].collider;
+        instances[instanceIdx].colliderCenter.x = entities[entIdx].colliderCenter.x;
+        instances[instanceIdx].colliderCenter.y = entities[entIdx].colliderCenter.y;
+        instances[instanceIdx].colliderCenter.z = entities[entIdx].colliderCenter.z;
+        instances[instanceIdx].colliderSize.x = entities[entIdx].colliderSize.x;
+        instances[instanceIdx].colliderSize.y = entities[entIdx].colliderSize.y;
+        instances[instanceIdx].colliderSize.z = entities[entIdx].colliderSize.z;
+        instances[instanceIdx].colliderMeshIndex = entities[entIdx].colliderMeshIndex;
+        instances[instanceIdx].mass = entities[entIdx].mass > 0.0f ? entities[entIdx].mass : 1.0f; // Nonzero fallback.
+        instances[instanceIdx].linearDrag = entities[entIdx].linearDrag > 0.0f ? entities[entIdx].linearDrag : 0.0f;
+        instances[instanceIdx].angularDrag = entities[entIdx].angularDrag > 0.0f ? entities[entIdx].angularDrag : 0.05f;
+        dirtyInstances[instanceIdx] = true;
+    }
     
     // Apply the Unity hierarchy nonsense, TODO: Save out level#.txt from the engine just once and then delete all this.
     if (levelCurrentlyLoading && entIdx != 755 && entIdx != 590) { // Adjusted for in the level data directly, no correction.
         instances[instanceIdx].position.x += correctionX;   
         instances[instanceIdx].position.y += correctionY;
         instances[instanceIdx].position.z += correctionZ;
-        if (ConstIndexIsDoor(entIdx)) {
+        if (isLight) {
+            instances[instanceIdx].position.x += correctionLightX;
+            instances[instanceIdx].position.y += correctionLightY;
+            instances[instanceIdx].position.z += correctionLightZ;
+        } else if (ConstIndexIsDoor(entIdx)) {
             instances[instanceIdx].position.x += correctionDoorsX;
             instances[instanceIdx].position.y += correctionDoorsY;
             instances[instanceIdx].position.z += correctionDoorsZ;
@@ -514,13 +565,8 @@ void AddInstance(uint16_t entIdx, uint16_t instanceIdx, uint32_t lineNum) {
             instances[instanceIdx].position.x += correctionNPCX;
             instances[instanceIdx].position.y += correctionNPCY - 1.0f; // Offset to center them up in their capsule
             instances[instanceIdx].position.z += correctionNPCZ;
-            
             Vector3 axis = (Vector3){0.0f, 0.0f, 1.0f}; // X-axis
             instances[instanceIdx].rotation = axis_angle_quaternion(axis, deg2rad(-90.0f));
-        } else if (ConstIndexIsLightStaticSaveable(entIdx)) {
-            instances[instanceIdx].position.x += correctionLightX;
-            instances[instanceIdx].position.y += correctionLightY;
-            instances[instanceIdx].position.z += correctionLightZ;
         } else if (ConstIndexIsStaticObjectSaveable(entIdx)) {
             instances[instanceIdx].position.x += correctionStaticSaveableX;
             instances[instanceIdx].position.y += correctionStaticSaveableY;
@@ -536,7 +582,6 @@ void AddInstance(uint16_t entIdx, uint16_t instanceIdx, uint32_t lineNum) {
         } 
     }
 
-    dirtyInstances[instanceIdx] = true;
     loadedInstances++;
 }
 
@@ -605,24 +650,23 @@ void LoadLevel(uint8_t curlevel) {
     GetLevel_NPCsSaveableInstantiated_ContainerOffsets(curlevel, &correctionNPCX, &correctionNPCY, &correctionNPCZ);
     while (fgets(lineSpace, lineLengthMax, file)) {
         size_t len = strlen(lineSpace);
-        while (len && (lineSpace[len - 1] == '\n' || lineSpace[len - 1] == '\r'))
-        lineSpace[--len] = '\0';
+        while (len && (lineSpace[len - 1] == '\n' || lineSpace[len - 1] == '\r')) lineSpace[--len] = '\0';
         line = lineSpace;
+        if (line[0] == '/' && line[1] == '/') continue;
+        
         snprintf(initialLine, sizeof(initialLine), "%s", line);
         memcpy(firstKeyCheck,line,10); firstKeyCheck[10] = '\0';
         lineNum++;
+        instanceIdx++;
+        if (instanceIdx >= INSTANCE_COUNT) { DualLogError("Too many instances %u in level%d.txt!\n",instanceIdx,curlevel); exit(1); }
+        
         bool isLight = true;
         if (strcmp(firstKeyCheck, "constIndex") == 0) isLight = false;  // constIndex specified indicating this is a real entity?
-        if (isLight) {
+        if (isLight) {            
             lightsIdx++;
-            lightIsDynamic[lightsIdx] = false;
             if (lightsIdx >= LIGHT_COUNT) { DualLogError("Too many lights %u in level%d.txt!\n",lightsIdx,curlevel); exit(1); }
-        } else {
-            instanceIdx++;
-            if (instanceIdx >= INSTANCE_COUNT) { DualLogError("Too many instances %u in level%d.txt!\n",instanceIdx,curlevel); exit(1); }
         }
         
-        int32_t litIdx = lightsIdx * LIGHT_DATA_SIZE;
         uint8_t lightType = 0u; // Point
         while(line[0] != '\0') {
             // Guaranteed no leading whitespaces,k comments, or blank lines, so don't bother
@@ -653,64 +697,46 @@ void LoadLevel(uint8_t curlevel) {
             trimmed_value[sizeof(trimmed_value) - 1] = '\0';
 //             sanitize_utf8_ascii(trimmed_key);
 //             sanitize_utf8_ascii(trimmed_value);
-            if (isLight) {
-                     if (strcmp(trimmed_key, "localPosition.x") == 0) lights[litIdx + LIGHT_DATA_OFFSET_POSX] = parse_float(trimmed_value, initialLine, lineNum) + /*correctionX*/25.56f + correctionLightX;
-                else if (strcmp(trimmed_key, "localPosition.y") == 0) lights[litIdx + LIGHT_DATA_OFFSET_POSY] = parse_float(trimmed_value, initialLine, lineNum) + /*correctionY*/-48.64f + correctionLightY;
-                else if (strcmp(trimmed_key, "localPosition.z") == 0) lights[litIdx + LIGHT_DATA_OFFSET_POSZ] = parse_float(trimmed_value, initialLine, lineNum) + /*correctionZ*/-5.2f + correctionLightZ;
-                else if (strcmp(trimmed_key, "localRotation.x") == 0) lights[litIdx + LIGHT_DATA_OFFSET_SPOTDIRX] = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localRotation.y") == 0) lights[litIdx + LIGHT_DATA_OFFSET_SPOTDIRY] = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localRotation.z") == 0) lights[litIdx + LIGHT_DATA_OFFSET_SPOTDIRZ] = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localRotation.w") == 0) lights[litIdx + LIGHT_DATA_OFFSET_SPOTDIRW] = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "intensity") == 0)       lights[litIdx + LIGHT_DATA_OFFSET_INTENSITY] = parse_float(trimmed_value, initialLine, lineNum) * 0.4; // Adjustment, globally applied from Citadel's Unity to Custom Game Engine (Voxen) conversion.
-                else if (strcmp(trimmed_key, "range") == 0)           lights[litIdx + LIGHT_DATA_OFFSET_RANGE] = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "spotAngle") == 0)       lights[litIdx + LIGHT_DATA_OFFSET_SPOTANG] = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "type") == 0) {
-                    if ((strcmp(trimmed_value, "Spot") == 0)) lightType = 1u;
-                    else if ((strcmp(trimmed_value, "Directional") == 0)) lightType = 2u;
-                }
-                else if (strcmp(trimmed_key, "color.r") == 0)         lights[litIdx + LIGHT_DATA_OFFSET_R] = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "color.g") == 0)         lights[litIdx + LIGHT_DATA_OFFSET_G] = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "color.b") == 0)         lights[litIdx + LIGHT_DATA_OFFSET_B] = parse_float(trimmed_value, initialLine, lineNum);
-            } else {
-                     if (strcmp(trimmed_key, "constIndex") == 0)      instances[instanceIdx].index = parse_numberu16(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localPosition.x") == 0) instances[instanceIdx].position.x = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localPosition.y") == 0) instances[instanceIdx].position.y = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localPosition.z") == 0) instances[instanceIdx].position.z = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localRotation.x") == 0) instances[instanceIdx].rotation.x = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localRotation.y") == 0) instances[instanceIdx].rotation.y = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localRotation.z") == 0) instances[instanceIdx].rotation.z = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localRotation.w") == 0) instances[instanceIdx].rotation.w = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localScale.x") == 0)    instances[instanceIdx].scale.x = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localScale.y") == 0)    instances[instanceIdx].scale.y = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "localScale.z") == 0)    instances[instanceIdx].scale.z = parse_float(trimmed_value, initialLine, lineNum);
-                else if (strcmp(trimmed_key, "go.activeSelf") == 0)   flag_set(&instances[instanceIdx].entflags, ENTFLAG_ACTIVE, parse_bool(trimmed_value, initialLine, lineNum));
-            }
+                 if (strcmp(trimmed_key, "constIndex") == 0)      instances[instanceIdx].index = parse_numberu16(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localPosition.x") == 0) instances[instanceIdx].position.x = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localPosition.y") == 0) instances[instanceIdx].position.y = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localPosition.z") == 0) instances[instanceIdx].position.z = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localRotation.x") == 0) instances[instanceIdx].rotation.x = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localRotation.y") == 0) instances[instanceIdx].rotation.y = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localRotation.z") == 0) instances[instanceIdx].rotation.z = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localRotation.w") == 0) instances[instanceIdx].rotation.w = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localScale.x") == 0)    instances[instanceIdx].scale.x = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localScale.y") == 0)    instances[instanceIdx].scale.y = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "localScale.z") == 0)    instances[instanceIdx].scale.z = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "go.activeSelf") == 0)                flag_set(&instances[instanceIdx].entflags, ENTFLAG_ACTIVE, parse_bool(trimmed_value, initialLine, lineNum));
+            else if (strcmp(trimmed_key, "gameObject.activeInHierarchy") == 0) flag_set(&instances[instanceIdx].entflags, ENTFLAG_ACTIVE, parse_bool(trimmed_value, initialLine, lineNum));
+            else if (strcmp(trimmed_key, "intensity") == 0)       instances[instanceIdx].intensity =  parse_float(trimmed_value, initialLine, lineNum) * 0.4; // Adjustment, globally applied from Citadel's Unity to Custom Game Engine (Voxen) conversion.
+            else if (strcmp(trimmed_key, "range") == 0)           instances[instanceIdx].range =      parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "spotAngle") == 0)       instances[instanceIdx].spotAngle =  parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "type") == 0) { if ((strcmp(trimmed_value, "Spot") == 0)) {lightType = 1u; } else if ((strcmp(trimmed_value, "Directional") == 0)) { lightType = 2u; } } // Not using ternary operator ?: due to needing to preserve lightType at 0u for Point lights
+            else if (strcmp(trimmed_key, "color.r") == 0)         instances[instanceIdx].color.r = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "color.g") == 0)         instances[instanceIdx].color.g = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "color.b") == 0)         instances[instanceIdx].color.b = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "lightOn") == 0)         flag_set(&instances[instanceIdx].entflags, ENTFLAG_LIGHT_ON, parse_bool(trimmed_value, initialLine, lineNum));
+            else if (strcmp(trimmed_key, "lerpOn") == 0)          flag_set(&instances[instanceIdx].entflags, ENTFLAG_LERP_ON, parse_bool(trimmed_value, initialLine, lineNum));
+            else if (strcmp(trimmed_key, "currentStep") == 0)     instances[instanceIdx].currentStep = parse_numberu8(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "lerpValue") == 0)       instances[instanceIdx].lerpValue = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "lerpTime") == 0)        instances[instanceIdx].lerpTime = LoadRelativeTimeDifferential(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "stepTime") == 0)        instances[instanceIdx].stepTime = parse_float(trimmed_value, initialLine, lineNum);
+            else if (strcmp(trimmed_key, "lerpStartTime") == 0)   instances[instanceIdx].lerpStartTime = LoadRelativeTimeDifferential(trimmed_value, initialLine, lineNum);
         }
         
-        if (isLight) {
-            loadedLights++;
-            lightsRangeSquared[lightsIdx] = lights[litIdx + LIGHT_DATA_OFFSET_RANGE] * lights[litIdx + LIGHT_DATA_OFFSET_RANGE];
-            // TODO: Set lightIsDynamic[lightsIdx] = true when light has animation data values set from file
-            if (lightType == 1) {
-                if (lights[litIdx + LIGHT_DATA_OFFSET_SPOTANG] < 5.0f) DualLogWarn("Spotlight %d on line %d loaded with spotAngle less than 5deg\n",lightsIdx,lineNum+1);
-            } else if (lightType == 2) {
-                // TODO: Handle directional lights for cyberspace
-                lights[litIdx + LIGHT_DATA_OFFSET_SPOTANG] = 0.0f; // Force to not be a spot light
-            } else {
-                lights[litIdx + LIGHT_DATA_OFFSET_SPOTANG] = 0.0f; // Force to not be a spot light
-            }
-        } else {
-            uint16_t parent = instanceIdx; // Needed as adding children moves the instanceIdx.
-            uint16_t entIdx = instances[parent].index;
-            AddInstance(entIdx, parent, lineNum);
-            AddChild0(entities[entIdx].child0, parent, entIdx, &instanceIdx, lineNum);
-            AddChild1(entities[entIdx].child1, parent, entIdx, &instanceIdx, lineNum);
-        }
+        if (isLight) loadedLights++;
+        flag_set(&instances[instanceIdx].entflags, ENTFLAG_IS_LIGHT, isLight);
+        if (lightType != 1 && isLight) instances[instanceIdx].spotAngle = 0.0f; // Force to not be a spot light for Point and Directional
+        uint16_t parent = instanceIdx; // Needed as adding children moves the instanceIdx.
+        uint16_t entIdx = instances[parent].index;
+        AddInstance(entIdx, parent, lineNum);
+        if (entIdx != UINT16_MAX) AddChild0(entities[entIdx].child0, parent, entIdx, &instanceIdx, lineNum);
+        if (entIdx != UINT16_MAX) AddChild1(entities[entIdx].child1, parent, entIdx, &instanceIdx, lineNum);
     }
 
     fclose(file);
-    
-    // Set Fog
     switch(curlevel) {
         case  0: fogColorR = 0.3207547f;  fogColorG = 0.29200783f;  fogColorB = 0.29200783f;  fogBaseDensityForLevel = 0.07f;  break;
         case  1: fogColorR = 0.34509805f; fogColorG = 0.38431373f;  fogColorB = 0.49019608f;  fogBaseDensityForLevel = 0.055f; break;
@@ -745,9 +771,6 @@ void LoadLevel(uint8_t curlevel) {
     GLuint groupX_shadClear = (totalShadowmapPixels + 31) / 32;
     glDispatchCompute(groupX_shadClear,1, 1);
     glMemoryBarrier(GL_SHADER_IMAGE_ACCESS_BARRIER_BIT);
-    numDynamicLights = 0;
-    for (int i=0;i<loadedLights;++i) { if (lightIsDynamic[i]) numDynamicLights++; }
-    DualLog("%u dynamic lights in level %u\n", numDynamicLights, currentLevel);
     //play_mp3("./Audio/music/THM1-19_medicalstart.mp3",((float)settings_VolumeMusic/100.0f) * 0.4f,100);
     RenderShadowmaps();
     Input_MouselookApply();
diff --git a/entity.h b/entity.h
index 001544d..5e339dd 100644
--- a/entity.h
+++ b/entity.h
@@ -16,6 +16,9 @@
 #define ENTFLAG_RIGIDBODY    32
 #define ENTFLAG_DOUBLESIDED  64
 #define ENTFLAG_TRANSPARENT 128
+#define ENTFLAG_IS_LIGHT    256
+#define ENTFLAG_LIGHT_ON    512
+#define ENTFLAG_LERP_ON    1024
 
 typedef struct {
     uint16_t index; // constIndex for entity type, used for indexing into arrays for resourec types when loading resources
@@ -32,6 +35,16 @@ typedef struct {
     Quaternion rotation; // Orientation matching Unity convention
     Vector3 scale;
     
+    float intensity;
+    float range;
+    float spotAngle;
+    Color color;
+    uint8_t currentStep;
+    float lerpValue;
+    float lerpTime;
+    float stepTime;
+    float lerpStartTime;
+    
     Vector3 velocity;
     Vector3 angularVelocity;
     
@@ -84,6 +97,8 @@ bool parse_data_file(DataParser *parser, const char *filename);
 #define GEOMETRY_LOD_CARD_MODEL_IDX 178
 extern Entity* entities; // Global array of entity definitions
 extern Entity instances[INSTANCE_COUNT];
+extern uint16_t lightsIndirection[INSTANCE_COUNT];
+extern uint16_t firstRelevantLightsIndirectionIndex;
 extern uint16_t loadedInstances;
 extern float modelMatrices[INSTANCE_COUNT * 16];
 extern uint8_t dirtyInstances[INSTANCE_COUNT];
diff --git a/voxen b/voxen
index 83c900a..90c958c 100755
Binary files a/voxen and b/voxen differ
diff --git a/voxen.c b/voxen.c
index 440e3de..0ba9b5b 100644
--- a/voxen.c
+++ b/voxen.c
@@ -397,6 +397,34 @@ Quaternion cubemapOrientationQuaternion[6] = {
     {0.0f, 1.0f, 0.0f, 0.0f}                   // -Z: Backward
 };
 
+void UpdateDynamicLights() {
+    for (int i = firstRelevantLightsIndirectionIndex; i < loadedInstances; ++i) {
+        if (!(instances[i].entflags & ENTFLAG_IS_LIGHT)) continue;
+
+        int lightsIdx = lightsIndirection[i];
+        if (lightsIdx >= UINT16_MAX) continue;
+        
+        int litIdx = lightsIdx * LIGHT_DATA_SIZE;
+        lights[litIdx + LIGHT_DATA_OFFSET_POSX] = instances[i].position.x;
+        lights[litIdx + LIGHT_DATA_OFFSET_POSY] = instances[i].position.y;
+        lights[litIdx + LIGHT_DATA_OFFSET_POSZ] = instances[i].position.z;
+        lights[litIdx + LIGHT_DATA_OFFSET_INTENSITY] = (instances[i].entflags & ENTFLAG_LIGHT_ON) ? instances[i].intensity : 0.0f;
+        lights[litIdx + LIGHT_DATA_OFFSET_RANGE] = instances[i].range;
+        lights[litIdx + LIGHT_DATA_OFFSET_SPOTANG] = instances[i].spotAngle;
+        lights[litIdx + LIGHT_DATA_OFFSET_SPOTDIRX] = instances[i].rotation.x;
+        lights[litIdx + LIGHT_DATA_OFFSET_SPOTDIRY] = instances[i].rotation.y;
+        lights[litIdx + LIGHT_DATA_OFFSET_SPOTDIRZ] = instances[i].rotation.z;
+        lights[litIdx + LIGHT_DATA_OFFSET_SPOTDIRW] = instances[i].rotation.w;
+        lights[litIdx + LIGHT_DATA_OFFSET_R] = instances[i].color.r;
+        lights[litIdx + LIGHT_DATA_OFFSET_G] = instances[i].color.g;
+        lights[litIdx + LIGHT_DATA_OFFSET_B] = instances[i].color.b;
+        lightsRangeSquared[lightsIdx] = instances[i].range * instances[i].range;
+//         DualLog("LightsIdx %u intensity %f, range %f, pos %f %f %f, spotAngle %f, color %f %f %f\n", lightsIdx, instances[i].intensity, instances[i].range, instances[i].position.x, instances[i].position.y, instances[i].position.z, instances[i].spotAngle, instances[i].color.r, instances[i].color.g, instances[i].color.b);
+    }
+
+    glNamedBufferData(lightsID,loadedLights * LIGHT_DATA_SIZE * sizeof(float), lights, GL_DYNAMIC_DRAW);
+}
+
 void UpdateVoxelLightLists() {
     memset(voxelLightListsRaw, 0, VOXEL_COUNT * 4 * sizeof(uint32_t));
     memset(voxelLightListIndices, 0, VOXEL_COUNT * 2 * sizeof(uint32_t));
@@ -537,13 +565,14 @@ void VoxelLists() {
     voxelLightListsRaw = malloc(VOXEL_COUNT * 4 * sizeof(uint32_t));
     voxelLightListIndices = malloc(VOXEL_COUNT * 2 * sizeof(uint32_t));
     voxelLightListIndicesID = SetupSSBO(voxelLightListIndicesID, 26, VOXEL_COUNT * 2 * sizeof(uint32_t), NULL, GL_DYNAMIC_DRAW);
-    voxelLightListsRawID = SetupSSBO(voxelLightListsRawID, 27,  1008105 * sizeof(uint32_t), NULL, GL_DYNAMIC_DRAW);
+    voxelLightListsRawID = SetupSSBO(voxelLightListsRawID, 27,  VOXEL_COUNT * 4 * sizeof(uint32_t), NULL, GL_DYNAMIC_DRAW);
     shadowmapIndirectionList = malloc(loadedLights * sizeof(uint32_t));
     lightsID = SetupSSBO(lightsID, 19, loadedLights * LIGHT_DATA_SIZE * sizeof(float), NULL, GL_DYNAMIC_DRAW);
     lightView      = calloc(loadedLights * 6 * 4 * 4, sizeof(float));
     lightViewProj  = calloc(loadedLights * 6 * 4 * 4, sizeof(float));
     lightFrustumPlanes   = calloc(loadedLights * 6 * 6, sizeof(FrustumPlane));
     DebugRAM("prior to UpdateVoxelLightLists");
+    UpdateDynamicLights();
     UpdateVoxelLightLists();
     float mat[16] = {1,0,0,0, 0,1,0,0, 0,0,1,0, 0,0,0,1};
     memcpy(&modelMatrices[0], mat, 16 * sizeof(float)); // Null instance matrix used for UI
@@ -683,7 +712,6 @@ void RenderShadowmaps(void) {
     uint32_t numToRender = vmin(numShadowsCouldRender, MAX_SHADOWMAPS);
     for (uint32_t c = 0; c < numToRender; ++c) { // Render top MAX_SHADOWMAPS candidates
         uint16_t lightIdx = candidates[c].index;
-        uint16_t litIdx = lightIdx * LIGHT_DATA_SIZE;
         uint32_t slot = shadowDrawCallsRenderedThisFrame;
         shadowmapIndirectionList[lightIdx] = slot;
         RenderShadowmap(lightIdx);
@@ -1325,16 +1353,11 @@ int32_t main(int32_t argc, char* argv[]) {
             glBindBuffer(GL_SHADER_STORAGE_BUFFER, matricesBuffer);
             glBufferData(GL_SHADER_STORAGE_BUFFER, loadedInstances * 16 * sizeof(float), modelMatrices, GL_DYNAMIC_DRAW);
             
-            // 3. Dynamic Shadowmaps
+            // 3. Light Updates
+            UpdateDynamicLights();
             for (int i = 0; i < loadedLights; ++i) {
                 if (lightDirty[i]) {
                     UpdateVoxelLightLists(); // Takes 12ms of total frametime!!
-//                     if (settings_Shadows > 0u) RenderShadowmaps();
-//                     else {
-//                         memset(shadowmapIndirectionList, MAX_SHADOWMAPS + 1, loadedLights * sizeof(uint32_t));
-//                         glNamedBufferData(shadowMapsIndirectionID, loadedLights * sizeof(uint32_t), shadowmapIndirectionList, GL_DYNAMIC_DRAW);
-//                     }
-                    
                     break;
                 }
             }
@@ -1345,7 +1368,6 @@ int32_t main(int32_t argc, char* argv[]) {
                 memset(shadowmapIndirectionList, MAX_SHADOWMAPS + 1, loadedLights * sizeof(uint32_t));
                 glNamedBufferData(shadowMapsIndirectionID, loadedLights * sizeof(uint32_t), shadowmapIndirectionList, GL_DYNAMIC_DRAW);
             }
-            glNamedBufferData(lightsID,loadedLights * LIGHT_DATA_SIZE * sizeof(float), lights, GL_DYNAMIC_DRAW);
 
             // 4. Raterized Geometry, Standard vertex + fragment rendering, but with special packing to minimize transfer data amounts
             glBindFramebuffer(GL_FRAMEBUFFER, gBufferFBO);
@@ -1495,7 +1517,7 @@ int32_t main(int32_t argc, char* argv[]) {
         float debugTextStartY = GetScreenRelativeY(0.075f);
         float leftPad = GetScreenRelativeX(0.0125f);
         if (!noHUD && showLocation) RenderFormattedText(leftPad, debugTextStartY, UI_LAYER_1, TEXT_WHITE, FONT_NORMAL, "x: %.4f, y: %.4f, z: %.4f", instances[PLAYER1].position.x, instances[PLAYER1].position.y, instances[PLAYER1].position.z);
-        if (!noHUD) RenderFormattedText(leftPad, debugTextStartY + (lineSpacing * 1), UI_LAYER_1, TEXT_WHITE, FONT_NORMAL, "timeSinceLastPhysicsTick: %.6f, numShadowsCouldRender: %u, playerCellIdx: %u", timeSinceLastPhysicsTick, numShadowsCouldRender, playerCellIdx);
+        if (!noHUD) RenderFormattedText(leftPad, debugTextStartY + (lineSpacing * 1), UI_LAYER_1, TEXT_WHITE, FONT_NORMAL, "timeSinceLastPhysicsTick: %.6f, numShadowsCouldRender: %u, playerCellIdx: %u, numCellsVisible: %u", timeSinceLastPhysicsTick, numShadowsCouldRender, playerCellIdx, numCellsVisible);
         if (!noHUD) RenderFormattedText(leftPad, debugTextStartY + (lineSpacing * 2), UI_LAYER_1, TEXT_WHITE, FONT_NORMAL, "Player velocity: %.2f, %.2f, %.2f, accumulated force: %.2f, %.2f, %.2f", instances[PLAYER1].velocity.x, instances[PLAYER1].velocity.y, instances[PLAYER1].velocity.z, instances[PLAYER1].accumulatedForce.x, instances[PLAYER1].accumulatedForce.y, instances[PLAYER1].accumulatedForce.z);
         if (consoleActive) RenderFormattedText(leftPad, 0, UI_LAYER_1, TEXT_WHITE, FONT_NORMAL, "] %s",consoleEntryText);
         if (statusTextDecayFinished > current_time) RenderFormattedText(leftPad + (screen_width / 2) - 220, screenCenterY - GetScreenRelativeY(0.30f + (genericTextHeightFac * 2.0f)), UI_LAYER_1, TEXT_WHITE, FONT_NORMAL, "%s",statusText);
diff --git a/voxen.h b/voxen.h
index 6542bb3..7510d41 100644
--- a/voxen.h
+++ b/voxen.h
@@ -160,7 +160,6 @@ void LoadModels(void);
 
 extern float lights[LIGHT_COUNT * LIGHT_DATA_SIZE];
 extern float lightsRangeSquared[LIGHT_COUNT];
-extern bool lightIsDynamic[LIGHT_COUNT];
 extern GLuint shadowMapSSBO;
 void VoxelLists();
 extern GLuint shadowmapsClearShaderProgram;
@@ -193,7 +192,6 @@ extern bool levelCurrentlyLoading;
 #define VOXEL_SIZE 0.32f
 #define VOXEL_HALF (VOXEL_SIZE * 0.5f)
 #define CELL_SIZE 2.56f // Each cell is 2.56x2.56
-#define MAX_SQUARE_DIST_INT 268435456 // (64 * 256)^2 = max square dist
 #define MAX_LIGHTS_PER_VOXEL 24 // Cap to prevent overflow
 #define CELL_VISIBLE       1
 #define CELL_OPEN          2
@@ -203,10 +201,6 @@ extern bool levelCurrentlyLoading;
 #define CELL_CLOSEDWEST   32
 #define CELL_SEES_SUN     64
 #define CELL_SEES_SKYBOX 128
-extern uint16_t numberOfFOVConeChecks0; // 5049
-extern uint16_t numberOfFOVConeChecks1; // 5030
-extern uint16_t numberOfFOVConeChecks2; // 2687
-extern uint16_t numberOfFOVConeChecks3; // 2687
 extern uint16_t playerCellIdx, playerCellIdx_x, playerCellIdx_y, playerCellIdx_z;
 extern uint16_t numCellsVisible;
 extern uint8_t gridCellStates[ARRSIZE];
